### PR TITLE
Fixed #34629 -- Added filtering support to GIS aggregates.

### PR DESCRIPTION
**Pull request by olethanh**
This fix an AttributeError when trying to use a filter argument in django.contrib.gis.db.models.aggregates.Collect and other GeoAggregate

Unit test have been added to check that this is now fixed.

See for description https://code.djangoproject.com/ticket/34629#ticket

### CODE REVIEW

File: tests/gis_tests/relatedapp/tests.py [hunk 1]
```diff
@@ -304,6 +306,41 @@ def test_collect(self):
         self.assertEqual(4, len(coll))
         self.assertTrue(ref_geom.equals(coll))
 
+    @skipUnlessDBFeature("supports_collect_aggr")
+    def test_collect_filter(self):
+        """
+        Testing the `Collect` aggregate with a filter
+        """
```
Comment (shangxiao) [inline tests/gis_tests/relatedapp/tests.py:313]:
```suggestion
```
Docstrings & comments that describe the _what_ instead of _why_ are discouraged. The method name should be descriptive enough to understand the problem. 😊

Edit: stupid phone typos 😭

**Comment by olethanh** [comment]
@shangxiao Thanks, you are right, I integrated the modification

File: tests/gis_tests/relatedapp/tests.py [hunk 2]
```diff
@@ -304,6 +306,38 @@ def test_collect(self):
...
+        kecksburg = annotated_queryset.get(name="Kecksburg")
+        self.assertEqual(kecksburg.snack_points, None)
+        self.assertEqual(kecksburg.snack_center, None)
+        roswell = annotated_queryset.get(name="Roswell")
+        self.assertEqual(roswell.snack_points, None)
+        self.assertEqual(roswell.snack_center, None)
+        dallas = annotated_queryset.get(name="Dallas")
+        # Postgis keep a MULTIPOINT but Spatialite collapse it as a point.
+        self.assertIn(dallas.snack_points.wkt, ("MULTIPOINT (7 8)", "POINT (7 8)"))
+        self.assertEqual(dallas.snack_center.wkt, "POINT (7 8)")
```
Comment (felixxm) [inline tests/gis_tests/relatedapp/tests.py:339]: Do we need a new model and objects? We should try to reuse existing cities e.g.
```python
    @skipUnlessDBFeature("supports_collect_aggr")
    def test_collect_filter(self):
        ref_geom = GEOSGeometry(
            "MULTIPOINT(-97.516111 33.058333,-96.801611 32.782057,"
            "-95.363151 29.763374,-96.801611 32.782057)"
        )   
        qs = City.objects.aggregate(
            points=Collect("location__point", filter=Q(state="TX"))
        )   
        collected_points = qs["points"]
        self.assertEqual(len(collected_points), 4)
        self.assertIs(ref_geom.equals(collected_points), True)
```
Can we add small test methods for other GIS aggregates?

**Review by felixxm — COMMENTED** [review]
@olethanh Thanks for this patch :+1:

The `filter` argument has never been documented or tested for GIS aggregates, so I'd treat this as a new feature, please add a release note (`5.0.txt`) and `filter=None` to the signatures of GIS [aggregates](https://docs.djangoproject.com/en/dev/ref/contrib/gis/geoquerysets/#aggregate-functions) with `versionchanged` annotation, e.g.

```
.. versionchanged:: 5.0

    Support for using the `filter` argument was added.
```

**Comment by felixxm** [comment]
@olethanh  Do you have time to keep working on this?

**Comment by olethanh** [comment]
yes sorry last week was quite busy but I should have sometime this afternoon and tomorrow

File: tests/gis_tests/relatedapp/tests.py (hunk 2 above)
Comment (olethanh) [reply tests/gis_tests/relatedapp/tests.py:339]: There is only one Location per City so I didn't think that it made much sense for theses tests.

I  can try using Parcel but there it require lot of arguments and they are no fixtures for them.
Will add tests for the other aggregate

**Comment by olethanh** [comment]
Hello, thanks for you review and advice.  I think I corrected all that was suggested. Can I ask you to review again when the tests pass?

- Re based the branch on main
- Updated doc and release notes.
- removed the test using the Snack Model and the model itself
- Added tests for all GIS aggregate functions. I prefer to group different parameter variation but tell me if I should subdivide them in even smaller test functions.
- Still had to add a new Model to tests the 3D functions and fields.

I have a crash when testing with MakeLine and there is only one Point, will open a separate ticket for that

File: tests/gis_tests/relatedapp/fixtures/initial.json [hunk 3]
```diff
@@ -78,62 +78,110 @@
       "state": "TX",
       "location": 4
     }
- },
- {
+  },
```
Comment (felixxm) [inline tests/gis_tests/relatedapp/fixtures/initial.json:81]: Please revert unrelated indentation changes.

File: docs/releases/5.0.txt [hunk 4]
```diff
@@ -170,6 +170,10 @@ Minor features
   function returns a 2-dimensional point on the geometry that is closest to
   another geometry.
 
+* Added support for filter argument on
+  :doc:`GIS aggregates </ref/contrib/gis/geoquerysets>`.
+
+
```
Comment (felixxm) [inline docs/releases/5.0.txt:176]: Two blank lines are unnecessary
```suggestion
```

File: docs/ref/contrib/gis/geoquerysets.txt [hunk 5]
```diff
@@ -917,10 +925,16 @@ Example:
     >>> print(qs["poly__extent3d"])
     (-96.8016128540039, 29.7633724212646, 0, -95.3631439208984, 32.782058715820, 0)
 
+.. versionchanged:: 5.0
+
+    Support for using the ``filter`` argument was added.
+
+
+
```
Comment (felixxm) [inline docs/ref/contrib/gis/geoquerysets.txt:933]:
```suggestion

```

File: django/contrib/gis/db/models/aggregates.py [hunk 6]
```diff
@@ -49,8 +49,8 @@ def resolve_expression(
         self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False
     ):
         c = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
-        for expr in c.get_source_expressions():
-            if not hasattr(expr.field, "geom_type"):
+        for field in c.get_source_fields():
+            if not hasattr(field, "geom_type"):
```
Comment (felixxm) [inline django/contrib/gis/db/models/aggregates.py:53]: Unfortunately, it's not enough to make `filter` works on Oracle, as [`tolerance`](https://github.com/django/django/blob/06881341d48dd17a4fdf25afc96991de4fb3b1ac/django/contrib/gis/db/models/aggregates.py#L34-L45) sneaks into `filter`:
```
======================================================================
ERROR: test_union_filter (gis_tests.relatedapp.tests.RelatedGeoModelTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/django/django/test/testcases.py", line 1428, in skip_wrapper
    return test_func(*args, **kwargs)
  File "/django/tests/gis_tests/relatedapp/tests.py", line 407, in test_union_filter
    city = qs.get(name="Aurora")
  File "/django/django/db/models/query.py", line 625, in get
    num = len(clone)
  File "/django/django/db/models/query.py", line 379, in __len__
    self._fetch_all()
  File "/django/django/db/models/query.py", line 1892, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/django/django/db/models/query.py", line 90, in __iter__
    results = compiler.execute_sql(
  File "/django/django/db/models/sql/compiler.py", line 1549, in execute_sql
    sql, params = self.as_sql()
  File "/django/django/db/models/sql/compiler.py", line 736, in as_sql
    extra_select, order_by, group_by = self.pre_sql_setup(
  File "/django/django/db/models/sql/compiler.py", line 84, in pre_sql_setup
    self.setup_query(with_col_aliases=with_col_aliases)
  File "/django/django/db/models/sql/compiler.py", line 73, in setup_query
    self.select, self.klass_info, self.annotation_col_map = self.get_select(
  File "/django/django/db/models/sql/compiler.py", line 296, in get_select
    sql, params = self.compile(col)
  File "/django/django/db/models/sql/compiler.py", line 544, in compile
    sql, params = vendor_impl(self, self.connection)
  File "/django/django/contrib/gis/db/models/aggregates.py", line 43, in as_oracle
    return clone.as_sql(
  File "/django/django/contrib/gis/db/models/aggregates.py", line 25, in as_sql
    return super().as_sql(
  File "/django/django/db/models/aggregates.py", line 138, in as_sql
    condition = When(self.filter, then=source_expressions[0])
  File "/django/django/db/models/expressions.py", line 1365, in __init__
    raise TypeError(
TypeError: When() supports a Q object, a boolean expression, or lookups as a condition.
```

File: docs/releases/5.0.txt [hunk 7]
```diff
@@ -170,6 +170,9 @@ Minor features
   function returns a 2-dimensional point on the geometry that is closest to
   another geometry.
 
+* Added support for filter argument on
+  :doc:`GIS aggregates </ref/contrib/gis/geoquerysets>`.
```
Comment (felixxm) [inline docs/releases/5.0.txt:174]: I would add an anchor and use it here:
```diff
diff --git a/docs/ref/contrib/gis/geoquerysets.txt b/docs/ref/contrib/gis/geoquerysets.txt
index eb2a29b900..1c98b6f896 100644
--- a/docs/ref/contrib/gis/geoquerysets.txt
+++ b/docs/ref/contrib/gis/geoquerysets.txt
@@ -839,6 +839,8 @@ Oracle      ``SDO_WITHIN_DISTANCE(poly, geom, 5)``
 SpatiaLite  ``PtDistWithin(poly, geom, 5)``
 ==========  ======================================
 
+.. _gis-aggregation-functions:
+
 Aggregate Functions
 -------------------
 

```

Comment (felixxm) [inline docs/releases/5.0.txt:174]:
```suggestion
* :ref:`GIS aggregates <gis-aggregation-functions>` now support the ``filter``
   argument.
```

**Review by felixxm — COMMENTED** [review]
@olethanh Thanks for updates :+1: I left more comments.

File: docs/releases/5.0.txt (hunk 7 above)
Comment (olethanh) [reply docs/releases/5.0.txt:174]: Thanks I couldn't figure out how to link to a particular  Heading in the document

File: django/contrib/gis/db/models/aggregates.py (hunk 6 above)
Comment (felixxm) [reply django/contrib/gis/db/models/aggregates.py:53]: > Unfortunately I don't have a Oracle on hand to test and there doesn't seem to be one in CI either.

There is an Oracle run in CI. It's called on demand, you need to leave a comment _"buildbot, test on oracle."_, see [CI wiki](https://code.djangoproject.com/wiki/CI).

**Comment by olethanh** [comment]
buildbot, test on oracle.

File: tests/gis_tests/relatedapp/tests.py [hunk 8]
```diff
@@ -304,6 +306,148 @@ def test_collect(self):
...
 
+    @skipUnlessDBFeature("supports_collect_aggr")
+    def test_collect_filter(self):
+        qs = City.objects.annotate(
+            parcel_centers=Collect(
+                "parcel__center1",
+                filter=~Q(parcel__name__icontains="ignore"),
+            )
+        )
+        city = qs.get(name="Aurora")
```
[the remark this answers is not in the document]
Comment (olethanh) [reply tests/gis_tests/relatedapp/tests.py]: you are right it make it nicer to read

**Comment by olethanh** [comment]
I've pushed a fix. Let's see if it resolves the issue. While the solution may not be the most elegant, I hope it helps prevent similar problems in the future. Personally, I find the code related to source expressions and filters quite confusing. The way filters are pushed and popped around can be challenging to follow."

**Comment by olethanh** [comment]
buildbot, test on oracle.

**Comment by olethanh** [comment]
Seems Oracle is failling with the error 
```
>    city = qs.get(name="Aurora")
< django.db.utils.DatabaseError: ORA-13050: unable to construct spatial object
```
in test_union_filter

https://djangoci.com/job/pull-requests-oracle/database=oragis19,label=oracle,python=python3.11/613/console

Full stacktrace
```

======================================================================
ERROR [0.053s]: test_union_filter (gis_tests.relatedapp.tests.RelatedGeoModelTest.test_union_filter)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/utils.py", line 98, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
cx_Oracle.DatabaseError: ORA-13050: unable to construct spatial object

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/test/testcases.py", line 1428, in skip_wrapper
    return test_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/tests/gis_tests/relatedapp/tests.py", line 404, in test_union_filter
    city = qs.get(name="Aurora")
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/models/query.py", line 625, in get
    num = len(clone)
          ^^^^^^^^^^
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/models/query.py", line 379, in __len__
    self._fetch_all()
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/models/query.py", line 1892, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/models/query.py", line 90, in __iter__
    results = compiler.execute_sql(
              ^^^^^^^^^^^^^^^^^^^^^
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/models/sql/compiler.py", line 1595, in execute_sql
    return list(result)
           ^^^^^^^^^^^^
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/models/sql/compiler.py", line 2096, in cursor_iter
    for rows in iter((lambda: cursor.fetchmany(itersize)), sentinel):
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/models/sql/compiler.py", line 2096, in <lambda>
    for rows in iter((lambda: cursor.fetchmany(itersize)), sentinel):
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/utils.py", line 97, in inner
    with self:
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/utils.py", line 91, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/home/jenkins/workspace/pull-requests-oracle/database/oragis19/label/oracle/python/python3.11/django/db/utils.py", line 98, in inner
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
django.db.utils.DatabaseError: ORA-13050: unable to construct spatial object

----------------------------------------------------------------------
```

I'm lacking  information for debug and I am not familira with Oracle, any idea?
Is it linked to this https://support.oracle.com/knowledge/Oracle%20Database%20Products/2564329_1.html ?

There is also an error in spatialite but it's unrelated to my change so I think it's a flaky test.

**Comment by olethanh** [comment]
buildbot, test on oracle.

I'm dividing the test in 3 so we can see if it is one of the annotate in particular that pause problem

**Comment by olethanh** [comment]
buildbot, test on oracle.

**Comment by olethanh** [comment]
I added a test without the filter, but it still fails on Oracle. Therefore, I suspect that the issue is not related to my recent change. It's possible that the problem lies with the test data I used.

I couldn't find any existing tests specifically for the Union Aggregate on Oracle. While there are passing tests for functions.Union in `geoapp/test_functions.py`, they are not related to the Aggregate functionality. I'm noticing a lot of workaround for Oracle in theses tests.

Is there a way to extract more details from Jenkins, such as the SQL queries executed? I'm feeling a bit lost on what steps to take next, so any guidance would be greatly appreciated.

I will be AFK for the next few days, but please feel free to provide any pointers or suggestions during my absence and i will continue looking into it upon my return.

**Comment by felixxm** [comment]
> I will be AFK for the next few days, but please feel free to provide any pointers or suggestions during my absence and i will continue looking into it upon my return.

Don't worry I will try to figure it out.

File: tests/gis_tests/relatedapp/tests.py [hunk 9]
```diff
@@ -304,6 +306,176 @@ def test_collect(self):
...
+        qs = City.objects.annotate(
+            parcel_border_union_single=Union(
+                "parcel__border2",
+                filter=Q(parcel__name__contains="Alpha"),
+            ),
+        )
+        city = qs.get(name="Aurora")
+        self.assertEqual(
+            city.parcel_border_union_single.wkt, "POLYGON ((0 0, 5 5, 8 8, 0 0))"
+        )
```
Comment (felixxm) [inline tests/gis_tests/relatedapp/tests.py:464]: These are already tested in `test_union_filter()`. Removed.

**Review by felixxm — COMMENTED** [review]
@olethanh Thanks :+1: I squashed commits, polished tests, and added tests for `Extent3D`.

Now I will check Oracle failures.

**Comment by felixxm** [comment]
> django.db.utils.DatabaseError: ORA-13050: unable to construct spatial object

Using `Union()` on provided polygons crashes even without filters, so I will switch `test_union_filter` to points. It's some kind of Oracle caveat that I wouldn't care about right now. My knowledge of Oracle GIS is quite limited.

I also fixed `GeoAggregate` on Oracle by pushing `SDOAGGRTYPE` wrapping to the `Func()`.

@olethanh Thanks again :+1:

**Comment by felixxm** [comment]
buildbot, test on oracle.

File: tests/gis_tests/relatedapp/tests.py (hunk 9 above)
Comment (olethanh) [reply tests/gis_tests/relatedapp/tests.py]: yes I had only split the test to check which annotate was failing

**Comment by olethanh** [comment]
@felixxm  Thanks a lot for the help, the fix and the merge !
